### DISCUSSION TITLE
Multi-ranger deck and motion capture system_crazyswarm

### DISCUSSION THREAD

**Discussion by MINJEEgi** [bitcraze/discussions#821] [Q&A] [answered]
Hello,
We are planning to map a space using multiple crazyflies with the multi-ranger deck through a motion capture system. Before conducting this experiment, we have a question and would like to reach out. We are wondering if it's possible to simultaneously receive values from multiple multi-ranger decks when using them in the motion capture system.

**Comment by ataffanel** [comment]
It depends where you want the multiple multi-ranger deck to be and where you want to get the values:

Connecting multiple multi-ranger deck to a single Crazyflie would be technically possible but will require hardware change to the deck (the I2C address of one chip needs to be changed) and to the driver. So this is possible but not easy.

Controlling multiple Crazyflie to one computer running Crazyswarm and getting ranging value from multiple Crazyflie in ROS should be no problem at all: you should be able to setup a log block to get the ranging from the Crazyflies and ROS is already setup to get logs from different Crazyflies (note that I have not used ROS or Crazyswarm for years so I cannot talk of the exact details, but from what I remember setting up custom log block resulted in one topic created for each connected Crazyflie generating the data of the log block).

**Reply by MINJEEgi** [reply]
Thank you for the response. So, just to clarify, I wanted to inquire if it is possible to connect one multi-ranger deck to one Crazyflie and simultaneously obtain values from each multi-ranger deck on the four Crazyflies. If we set up a log block, it seems like it should be possible, right?

**Reply by MINJEEgi** [reply]
Additionally, I have one more question. As mentioned earlier, when obtaining values from the four drones equipped with Multi-ranger decks, do we need to change the I2C addresses?

**Reply by ataffanel** [reply, marked as answer]
Yes it is possible to log the range from multiple Crazyflies in ROS and no there will be no I2C change in that case since the multi-ranger decks will be installed on different drones.

**Reply by MINJEEgi** [reply]
thank you for your response. :)

**Reply by MINJEEgi** [reply]
hello. While planning the process, I had a question, so I decided to ask again. 

1. Can we achieve precise control using optitrack tracking and receive data from the multi-ranger sensors (front, back, left, right)? We need two types of data: 'timestamp' and 'crazyflie 6-dof pose,' as well as corresponding 'timestamp' and 'distance measurements for front, back, left, and right.'

2. Does 'time won't be synchronized' mean that each drone operates on a different time reference point during their interaction? or related to the synchronization between the 6-dof pose and the multi-ranger values?

3. Also, we are curious if it is possible to use cflib instead of crazyswarm for our project.

**Reply by ataffanel** [reply]
Hi,

 1. Yes in principle you can get optitrack tracking at the same time as getting the ranges. The time-stamps might be an issue thought:
 2. From the Crazyflie logs you are getting the number of milliseconds since the Crazyflie has been started. There is currently no implemented way to synchronize Crazyflie clocks so the best might be to timestamp the arrival of the packets in the computer: as long as the links are not overloaded, the latency should not be higher than a couple of 10s of milliseconds. It would be technically possible to time synchronize Crazyflies over radio but this is not something that is currently implemented, I would be glad to point the direction if this is something you would like to look at though. Another point about time is that the Crazyflie's STM32 is running on a ceramic oscillator that only has +/- 0.5% of accuracy.
 3. In principle it is possible to use cflib, the difficulty is with the mocap system though: Crazyswarm is optimized to feed mocap position estimate very efficiently to the Crazyflies, this would need to be implemented manually with CFlib and so it will not work as well.
